x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/SnapshotLifecyclePolicy.java

```diff
@@ -81,14 +88,16 @@ public class SnapshotLifecyclePolicy implements SimpleDiffable<SnapshotLifecycle
     public SnapshotLifecyclePolicy(
         final String id,
         final String name,
-        final String schedule,
+        @Nullable final String schedule,
+        @Nullable final String interval,
         final String repository,
         @Nullable final Map<String, Object> configuration,
         @Nullable final SnapshotRetentionConfiguration retentionPolicy
     ) {
         this.id = Objects.requireNonNull(id, "policy id is required");
         this.name = Objects.requireNonNull(name, "policy snapshot name is required");
-        this.schedule = Objects.requireNonNull(schedule, "policy schedule is required");
+        this.schedule = schedule;
+        this.interval = interval;
```

Comment:
I would like this a bit better if there were validation in the constructor here that makes sure these cannot both be `null`. The previous logic required at this point in the constructor that the `schedule` wasn't `null`, and the validation that it was not an empty String was deferred to the `validate` method. I'd like to see something along the same lines, but handling both `interval` and `schedule`.

Reply:
I want your opinion this my change for this. Probably easiest to look at the commit diff: https://github.com/elastic/elasticsearch/pull/110847/commits/1fa9058464fa4100c6471fbc9a4b4eae2e66fc24

I would like to validate that at least one of schedule and interval is defined, and that at most one of schedule and interval is defined. It turns out that the parser provides some helpers for this, along with nice error messages: 
```
PARSER.declareExclusiveFieldSet(SCHEDULE.getPreferredName(), INTERVAL.getPreferredName());
PARSER.declareRequiredFieldSet(SCHEDULE.getPreferredName(), INTERVAL.getPreferredName());
```

Unfortunately, these interact rather trickily with validation in the constructor. 
The assertion in the constructor:
```
assert schedule != null || interval != null : "policy schedule or interval is required";
```
does the same thing as `declareRequiredFieldSet`. It's still nice to have the assertion in case someone calls the constructor directly. This works fine.

But if we add a similar assertion for the `at most one is defined case`:
```
assert schedule == null || interval == null : "only one of schedule or interval may be defined";
```
The assertion bubbles up to the caller. Not the error message provided by the parser. It seems that the constructor is called eagerly during parsing. For this reason I haven't included this assertion in the constructor, because I'd like to use the error message from the parser. 

Anyway, I'm just not super happy with the inconsistency between the two cases. If you have any ideas, I'm all ears.